### PR TITLE
Add assertions for PrestoException and SemanticException

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        extends AbstractThrowableAssert<PrestoExceptionAssert, PrestoException>
+{
+    @CheckReturnValue
+    public static PrestoExceptionAssert assertPrestoExceptionThrownBy(ThrowingCallable throwingCallable)
+    {
+        Throwable throwable = catchThrowable(throwingCallable);
+        if (throwable == null) {
+            throw Failures.instance().failure("Expecting code to raise a PrestoException");
+        }
+        assertThat(throwable).isInstanceOf(PrestoException.class);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java]: Will this propagate `throwable` (as cause / suppressed) when it's not a PE?

File: presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    private PrestoExceptionAssert(PrestoException actual)
+    {
+        super(actual, PrestoExceptionAssert.class);
+    }
+
+    public PrestoExceptionAssert hasErrorCode(ErrorCodeSupplier errorCodeSupplier)
+    {
+        assertThat(actual.getErrorCode()).isEqualTo(errorCodeSupplier.toErrorCode());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java]: When this fails, `actual` should be included (as cause / suppressed) in the AssertionError being raised.

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java]: (same in SEA)

File: presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java [hunk 3]
```diff
@@ -324,132 +323,55 @@ public void assertInvalidFunction(String projection)
...
-                assertTrue(e.getMessage().equals(messagePattern) || e.getMessage().matches(messagePattern), format("Error message [%s] doesn't match [%s]", e.getMessage(), messagePattern));
-            }
-            catch (Throwable failure) {
-                failure.addSuppressed(e);
-                throw failure;
-            }
-        }
+        assertPrestoExceptionThrownBy(() -> evaluateInvalid(projection))
+                .hasErrorCode(errorCode)
+                .hasMessage(message);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java]: Previously, the code allowed equality or pattern matching for error messages.
Now, with the fluent API allows us to treat them separately, but for callers of `assertInvalidFunction`, this is not a drop-in replacement.

File: presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java (hunk 1 above)
Comment (electrum) [reply presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java]: Yes
```java
assertThat(new RuntimeException("hello")).isInstanceOf(IOException.class);
```
produces
```
java.lang.AssertionError: 
Expecting:
  <java.lang.RuntimeException: hello>
to be an instance of:
  <java.io.IOException>
but was:
  <"java.lang.RuntimeException: hello
	at javatest11.AssertJ.assertException(AssertJ.java:29)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:73)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
">

```

File: presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java (hunk 3 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java]: Yes, based on my (somewhat incomplete) visual verification, all of the callers used messages. I renamed all of the parameters to reflect that.

I think that the previous behavior of allowing both is not a good API design.

File: presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java (hunk 2 above)
Comment (electrum) [reply presto-main/src/main/java/io/prestosql/testing/assertions/PrestoExceptionAssert.java]: Done

**Pull request merged**